bundles/automation/org.eclipse.smarthome.automation.module.script.rulesupport/src/main/java/org/eclipse/smarthome/automation/module/script/rulesupport/internal/ScriptedCustomModuleHandlerFactory.java

```diff
@@ -0,0 +1,87 @@
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ */
+package org.eclipse.smarthome.automation.module.script.rulesupport.internal;
+
+import java.util.Collection;
+import java.util.Dictionary;
+import java.util.HashMap;
+import java.util.Hashtable;
+
+import org.eclipse.smarthome.automation.Module;
+import org.eclipse.smarthome.automation.handler.ModuleHandler;
+import org.eclipse.smarthome.automation.handler.ModuleHandlerFactory;
+import org.eclipse.smarthome.automation.module.script.rulesupport.shared.ScriptedHandler;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceRegistration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link ScriptedCustomModuleHandlerFactory} is used in combination with the
+ * {@link ScriptedCustomModuleTypeProvider} to allow scripts to define custom types in the RuleEngine. These
+ * registered types can then be used publicly from any Rule-Editor.
+ *
+ * This class provides the handlers from the script to the RuleEngine. As Jsr223 languages have different needs, it
+ * allows these handlers to be defined in different ways.
+ *
+ * @author Simon Merschjohann
+ *
+ */
+public class ScriptedCustomModuleHandlerFactory extends AbstractScriptedModuleHandlerFactory {
+    private Logger logger = LoggerFactory.getLogger(ScriptedCustomModuleHandlerFactory.class);
+
+    private final HashMap<String, ScriptedHandler> typesHandlers = new HashMap<>();
+
+    private ServiceRegistration<?> bmhfReg;
+
+    private static ScriptedCustomModuleHandlerFactory instance;
+
+    public static ScriptedCustomModuleHandlerFactory get() {
+        return instance;
+    }
+
+    @Override
+    public void activate(BundleContext bundleContext) {
+        super.activate(bundleContext);
+
+        bmhfReg = bundleContext.registerService(ModuleHandlerFactory.class.getName(), this, null);
+
+        instance = this;
+    }
+
+    @Override
+    public Collection<String> getTypes() {
+        return typesHandlers.keySet();
+    }
+
+    @Override
+    protected ModuleHandler internalCreate(Module module, String ruleUID) {
+        ScriptedHandler scriptedHandler = typesHandlers.get(module.getTypeUID());
+
+        return getModuleHandler(module, scriptedHandler);
+    }
+
+    public void addModuleHandler(String uid, ScriptedHandler scriptedHandler) {
+        typesHandlers.put(uid, scriptedHandler);
+
+        updateModuleHandlerOffer();
+    }
+
+    public void removeModuleHandler(String uid) {
+        typesHandlers.remove(uid);
+
+        updateModuleHandlerOffer();
+    }
+
+    private void updateModuleHandlerOffer() {
+        Dictionary<String, Object> properties = new Hashtable<String, Object>();
+
+        properties.put("handlers", typesHandlers.values());
```

Comment:
> Should I look at that?

Would be wonderful :-)

> And should I integrate it in this PR or should I add another PR for that?

Probably best to do it as a separate PR, which reverts parts of your previous PR that had introduced that.
Although I think we are very soon ready to merge #1783 as well, it will have to go through the Eclipse IP process, so it will take slightly longer.